Feature dynamic vegetation

N1ckP3rsl3y opened the pull request:
- Merge dynamic vegetation calculation changes into working parallel branch to avoid more complicated conflicts in the future
- Add max depth soil temperature method to parallel mode
- Add dynamic vegetation calculation capabilities to parallel mode/working parallel branch

N1ckP3rsl3y committed:
New vegetation inputs - short- & long-term vegetation length

- This is the beginning commit to implement a new way to calculate/adjust vegetation being used by the program
- This will be dynamic vegetation and will adjust throughout the years (before every year) based on predictor variables

- veg.in gains new input options
	* nYearsDynamicLong - Number of years long-term vegetation predictors are summarized
	* nYearsDynamicShort - Number of years short-term vegetation predictors are summarized (as anomaly to long-term predictors)
	* (Modified) vegetation parameter (veg method) - New option for dynamic vegetation using the two variables above

SW_VEGPROD_INPUTS
	* Gain variables for nYearsDynamicShort & nYearsDynamicLong

SW_MPI.c
- Spread values of nYearsDynamicShort & nYearsDynamicLong across all ranks

`SW_VPD_read()`
- Adjust line numbers and read in new inputs of `veg.in`
- Error if any of the following occur
	* short-term value >= long-term
	* short- or long-term value is <= 0
	* veg_method is not 0, 1 or 2

N1ckP3rsl3y committed:
New vegetation inputs - short- & long-term vegetation length

- This is the beginning commit to implement a new way to calculate/adjust vegetation being used by the program
- This will be dynamic vegetation and will adjust throughout the years (before every year) based on predictor variables

- veg.in gains new input options
	* nYearsDynamicLong - Number of years long-term vegetation predictors are summarized
	* nYearsDynamicShort - Number of years short-term vegetation predictors are summarized (as anomaly to long-term predictors)
	* (Modified) vegetation parameter (veg method) - New option for dynamic vegetation using the two variables above

SW_VEGPROD_INPUTS
- Gain variables for nYearsDynamicShort & nYearsDynamicLong

SW_MPI.c
- Spread values of nYearsDynamicShort & nYearsDynamicLong across all ranks

`SW_VPD_read()`
- Adjust line numbers and read in new inputs of `veg.in`
- Error if any of the following occur
	* short-term value >= long-term
	* short- or long-term value is <= 0
	* veg_method is not 0, 1 or 2

dschlaep committed:
Derived metrics as new output

- close #466 "Derived metrics as output"

- Derived metrics including can now be requested as output
- Currently implemented: climatic water deficit, dry degree-days, wet degree-days, and total profile available soil moisture
- two new output groups: DERIVEDSUM, DERIVEDAVG

- Implementation
* new module SW_DerivedMetrics: functions that calculate derived metrics
* struct SW_SITE_SIM has new variables that hold intermediate, time-invariant variables (that are used to calculate derived metrics)
* SW_SIT_init_run() now populates the intermediate variables in SW_SITE_SIM
* new soilLayerWeight() to calculate the proportional contribution of the thickness of a soil layer to a depth limit
* updated output to incorporate the two new output groups
* new get_derivedsum_*() and get_derivedavg_*() functions to format the output
* new unit tests that cover new soilLayerWeight() and derived metrics functions

N1ckP3rsl3y committed:
Calculate constant soil info

- Part of the dynamic vegetation calculations will be constant values we can calculate before the simulation run occurs

- Add these calculations within `SW_VPD_init_run()`
- New struct SW_SOIL_SIM (put in SW_RUN) that holds the following information
	* Maximum soil depth
	* % sand within soil profile (weighted by the layer width)
	* % coarse fragments within soil profile (weighted by the layer width)
	* Total available water holding capacity
	* % clay within the first 3cm (or first layer, whichever is deeper) * % organic matter within the first 3cm (or first layer, whichever is deeper)

- `SW_VPD_init_run()` now handles the `veg_method` option of 2
	* This function now accepts and instance of SW_RUN rather than the previous 4+ structs
- Update call in `SW_CTL_init_run()`
- Create helper functions for `SW_VPD_init_run()` to calculate the constant values listed above
- `calc_const_dynamic_veg_info()` - wrapper function to calculate these constant values
- `calc_perc_var_in_soil_profile()`
	* Has two modes - "first3cm" enabled or disabled to switch between % in the first 3cm (or first layer) and the whole soil profile
- `calc_awhc()` - calculates available water holding capacity

- Update any tests to match the update to the interface of `SW_VPD_init_run()`

N1ckP3rsl3y committed:
Add new calculation/predictor variables for dynamic veg

- Dynamic vegetation calculations need further information before being able to actually calculate updated vegetation values
- This information includes
	- Temperature, precipitation, water deficits, isothermality and temp-precip correlations
	- This information is to be kept track for every year and will take short- and long-term averages plus anomalies (including % anomalies)

- SW_VEGPROD_SIM gains new variables to hold (arrays holding values for every year)
	* Annual monthly mean temperature
	* Correlation between each month's average temperature and precipitation
	* Annual isothermality
	* Annual water deficit
	* Annual precipiation
	* Annual seasonal precipitation
	* Annual precipitation of the driest month
	* Annual wet degree days
	* Maximum temperature of the hottest month for each year
	* Minimum temperature of the hottest month for each year
	* Precipitation of the wettest month of the year
- SW_VEGPROD_SIM also gains variables to hold short- and long-term averages of variables that will be used in future calculations
- SW_VEGPROD_SIM also gains variables to hold anomaly and anomaly % for variables that will be used in future calculations

- New functions - `SW_VPD_init_ptrs()` & `SW_VPD_deconstruct()` - to initialize dynamic pointers to NULL and free them if allocated
- New static function - `alloc_nyear_arrays()` (SW_VegProd.c) - will allocate the dynamic arrays to be of size <n years>

N1ckP3rsl3y committed:
New `veg_method` constants

- When comparing `veg_method` to get what to do with vegetation, 0, 1 and now 2 will become difficult to understand as a literal

- SW_VegProd.h gains new constants
	* VEG_METHOD_FILE (commented out, unused)
	* VEG_METHOD_LONG_EST (1) to specify the option to only use long-term climate conditions
	* VEG_METHOD_DYN_EST (2) to specify the option to use long- and short-term (dynamic) climate conditions

N1ckP3rsl3y committed:
Merge branch 'feature_parallel_parallelization' into feature_dynamic_vegetation

N1ckP3rsl3y committed:
Replace `<n years>` with `n years` because of `make doc`

N1ckP3rsl3y committed:
Rename `percClay` & `percOM` to `surface*`

- Reword the prefix "perc" to no longer represent the percentage but to explain that it is clay and OM at the surface, not through the soil profile

N1ckP3rsl3y committed:
Calculate annual values for predictor values

- Step 1 of 3 to calculating updated vegetation values
- Update the annual values for the newest year for the arrays in `SW_VEGPROD_SIM`

SW_VegProd.c
- New function `calc_yearly_hist_vals()` which calculates a current year's values for later calculations
- New function `update_veg_yearly()` , a wrapper function to fully update yearly vegetation values
- Update `SW_VPD_new_year()` interface to take in yearly weather, SW_VEGPROD_SIM, vegetation method, and start year of simulation
- `SW_VPD_new_year()` now check if vegetation is to be estimated and vegetation method is 2, call new function `update_veg_yearly()`

generic.c
- New function - `correlation_coefficient()` to calculate the correlation coefficient between two sets of values

SW_datastructs.h
- Make `estVeg` a variable inside SW_VEGPROD_SIM to have the ability to be sent around easier

N1ckP3rsl3y committed:
Calculate predictor variables for vegetation calculations

- Step 2 of 3 of calculating updated vegetation values
- Using values from the annual calculations, calculate
	* Long-term averages
	* Short-term averages
	* Anomaly
	* Rate of anomaly
based on said annual calculations for certain variables

SW_VegProd.c
- New function - `calc_veg_predictor_vals()` to calculate the predictor variables which will be used in updated vegetation values
- `SW_VPD_new_year()` & `update_veg_yearly()` gain new parameters to specify the short- and long-term lengths

N1ckP3rsl3y committed:
Calculate updated vegetation values

- Step 3 of 3 to calculate updated vegetation values for the year

SW_VegProd.c
- New function `calc_CONUS_vegcov_2025()` to calculate updated vegetation values
	* Currently is not hooked up to vegetation arrays due to other necessary changes but will be hooked up later
- `SW_VPD_new_year()` & `update_veg_yearly()` gain parameter to accept SW_SOIL_SIMS
- Rename `SW_WeathHist` to `SW_YearWeathHist` in header of `SW_VPD_new_year()` (missed in a previous commit)

N1ckP3rsl3y committed:
Fix descriptions of `nYearsDynamic*` in SW_VEGPROD_INPUTS

N1ckP3rsl3y committed:
Replace "anomaly %" with "rate of anomaly"

- Wording the rate of anomaly as "anomaly %" isn't the most accurate, if not incorrect

- Replace variable prefixes "anomPct" with "rateAnom"

N1ckP3rsl3y committed:
Merge branch 'feature_parallel_parallelization' into feature_dynamic_vegetation

N1ckP3rsl3y committed:
Help smooth shutdown upon early exit in sim

- Currently, when given an interrupt of sorts when simulations are running, shutdown does not occur
- The main problem being that the I/O processes do not receive the signals

SW_Control.c
- Move `signal()` calls to within `SW_CTL_RunSims()` so both I/O and compute processes can receive the signal
- Compute processes now expect inputs if the `runSims` flag is true

SW_MPI.c
- Only read inputs if `runSims` flag is true
- After the main loop of I/O processes, only write outputs if the `runSims` flag was set to false before reading inputs

Notes
- On an HPC environment, this for the most part works properly
- The only trouble is the program is getting stuck on closing input files and it has yet to be determined if there is anything wrong with the programs code
	* This happens most of the time but can also be inconsistent

N1ckP3rsl3y committed:
Merge branch 'feature_parallel_parallelization' into feature_dynamic_vegetation

N1ckP3rsl3y committed:
Create basic tests for vegetation calculations

- The code for the newest vegetation estimations has been split up and is easy to create unit tests for them

- Create 3 unit test suites that test the following functions
	* calc_yearly_hist_vals
	* calc_veg_predictor_vals
	* calc_perc_var_in_soil_profile
	* calc_awhc
	* correlation_coefficient

- New testing helper function - calc_veg_average - to switch between/calculate running and moving window averages

N1ckP3rsl3y committed:
New input in `siteparam.in` - Max depth soil temperature

- See issue https://github.com/DrylandEcology/SOILWAT2/issues/446
- New input to specify the method of determining soil temperature at maximum depth
	* 0 - User provided, static value
	* 1 - Long-term average of mean daily air temperature

- siteparam.in gains this new parameter
- Update SW_SIT_read() to properly read/set this value
- Update MPI input distribution (`SW_MPI_template_info()`) to spread this input across processes

N1ckP3rsl3y committed:
Provide functionality to new input (max depth soil temp)

- The program currently only reads in the new input to specify the method of determining the soil temperature at max depth
- Provide functionality to update the soil temperature constant for every year

- New function - `SW_SIT_new_year()` to update the soil temperature value
- Update `SW_VPD_init_run()` to allocate only annual mean daily temperature if the user only requested the new soil temp method == 1
- Update vegetation functions called within `SW_VPD_new_year()` to only calculate annual daily temperature
	* Calculate annual daily mean temperature if we do not estimate vegetation (or veg_method != 2) and the new soil temp method == 1

- Update tests to match the new interfaces of vegetation functions

- Remove <> around "<n layers>" which caused a problem with doxygen

N1ckP3rsl3y committed:
Merge branch 'feature_parallel_parallelization' into feature_dynamic_vegetation

dschlaep committed:
Merge branch 'feature_parallel_parallelization' into feature_dynamic_vegetation

* SW_SIT_read() now with correct value for nLinesWithoutTR (fixing commit 2be063b981da21d4cf3a894b485baf0c238c4d1b "New input in `siteparam.in` - Max depth soil temperature")

# Conflicts:
#	src/SW_Site.c

dschlaep committed:
Merge branch 'feature_parallel_parallelization' into feature_dynamic_vegetation

* Remove obsolete "estVeg"

# Conflicts:
# include/SW_VegProd.h
# include/SW_datastructs.h
# src/SW_Control.c
# src/SW_VegProd.c
# tests/gtests/test_SW_VegProd.cc
# tests/gtests/test_WaterBalance.cc

dschlaep committed:
Fix maxDepthSoilTemperature

- see issue #446
- commit e0fa632a15ca94ad12376f3e1c992889a3f53057 "Provide functionality to new input (max depth soil temp)" introduced the functionality to update maxDepthSoilTemperature every year

- SW_VPD_new_year() now uses a correct logic in SW_VPD_new_year() to calculate the value of long-term mean annual temperature even if dynamic vegetation is turned off (previously, it was stuck at 0)
- calc_veg_predictor_vals() now increments shortIndex based on nYearsDynamicShort and longIndex based on nYearsDynamicLong (previously, it was incrementing both relative to termLength which ended up being nYearsDynamicShort)
- new calc_annTempLongAvg() that calculates long-term mean annual temperature that is used if dynamic vegetation is turned off -- if dynamic vegetation is turned on, then all vegetation predictors, including long-term mean annual temperature, are calculated by calc_veg_predictor_vals()
- update_veg_yearly() now calls calc_annTempLongAvg() if maxDepthSoilTemperature but not vegetation predictors are requested -- previously, annual values but not long-term mean annual temperature was calculated if maxDepthSoilTemperature but not vegetation predictors were requested (which caused that maxDepthSoilTemperature was stuck at 0)
- SW_VPD_init_run() now allocates memory for annual temperature values used for maxDepthSoilTemperature also if veg_method is 1 (static vegetation cover estimated from simulation-wide climate)
- converted nYearsDynamicShort and nYearsDynamicLong from int to TimeInt (unsigned int) for easier comparisons against other TimeInt variables
-

- revert siteparam.in setting for maxDepthSoilTemperature to previous default (to facilitate comparison of output values with previous versions)

- addressed 'error: narrowing conversion from 'unsigned int' to signed type 'int' is implementation-defined [bugprone-narrowing-conversions,cppcoreguidelines-narrowing-conversions,-warnings-as-errors]'

dschlaep committed:
Merge branch 'feature_parallel_parallelization' into feature_dynamic_vegetation

dschlaep committed:
Merge branch 'feature_parallel_parallelization' into feature_dynamic_vegetation

- this commit includes v8.1.2

# Conflicts:
#	include/SW_datastructs.h
#	tests/gtests/test_SW_VegProd.cc

dschlaep committed:
Fix test VegProdTest.VegProdConstructor

- fix commit 2d55c2a52f6a2c33437bd1a9d252b3961d25a7f7: test now provides (again) values for variables utilized by SW_VPD_init_run()
- the above commit caused that these variables no longer contained values, and, because they were uninitialized, could take invalid values (e.g., see error below due to isnorth)
```
[ RUN      ] VegProdFixtureTest.VegProdConstructor
src/SW_VegProd.c:1806:44: runtime error: load of value 2863311530, which is not a valid value for type 'Bool'
SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior src/SW_VegProd.c:1806:44
```

- test VegProdConstructor no longer is a test fixture because it was not utilized

N1ckP3rsl3y committed:
Merge branch 'feature_parallel_simplify' into feature_dynamic_vegetation

N1ckP3rsl3y committed:
Merge branch 'feature_parallel_simplify' into feature_dynamic_vegetation

dschlaep committed:
calc_CONUS_vegcov_2025() now returns array of calculated vegetation covers

dschlaep reviewed (commented):
When I turn on dynamic vegetation, then I get a segfault when run with `bin_debug` and an unknown address error when run with `bin_debug_severe` or `bin_sanitizer`

```
==9558==ERROR: AddressSanitizer: BUS on unknown address (pc 0x00010493cab4 bp 0x00016b38db20 sp 0x00016b38d7a0 T0)
==9558==The signal is caused by a READ memory access.
==9558==Hint: this fault was caused by a dereference of a high value address (see register values below).  Disassemble the provided pc to learn which register was used.
    #0 0x10493cab4 in calc_veg_predictor_vals SW_VegProd.c:666
    #1 0x10493fd18 in update_veg_yearly SW_VegProd.c:1169
    #2 0x104947cac in SW_VPD_new_year SW_VegProd.c:2036
    #3 0x104893b78 in SW_CTL_run_current_year SW_Control.c:1110
    #4 0x1048935ec in SW_CTL_main SW_Control.c:464
    #5 0x10489702c in SW_CTL_run_sw SW_Control.c:1741
    #6 0x104896490 in SW_CTL_RunSimSet SW_Control.c:706
    #7 0x104884e38 in main SW_Main.c:200
    #8 0x194cb6b94  (<unknown module>)
```

dschlaep committed:
New tests for dynamic vegetation cover and dynamic soil temperature boundary condition

- new test WaterBalanceFixtureTest.WaterBalanceWithDynamicSoilTemperatureBoundary activates the dynamic soil temperature boundary condition

- new test WaterBalanceFixtureTest.WaterBalanceWithVegetationFromClimate2 activates the dynamic vegetation

- both use 3 years for short-term and 10 years for long-term conditions; these are shorter periods than the 31 simulated years to trigger the moving window

N1ckP3rsl3y committed:
Correct correlation coefficient calculation

- Update `correlation_coefficient` to take the sqrt() of the final calculation of the function

N1ckP3rsl3y committed:
Fix segmentation fault when running `bin_sanitizer`

- Currently a segmentation fault occurs when running `make CC=clang bin_sanitizer` while calculating dynamic vegetation values

- Initialize shortIndex/longIndex within SW_VEGPROD_SIM when initializing the simulation run

- Result: Program no longer throws a segmentation fault

N1ckP3rsl3y committed:
Correct dynamic vegetation calculations part 1

- The resulting values from `calc_CONUS_vegcov_2025()` do not sum to 1
- When debugging a few things have stood out to be incorrect

- No longer include current year in short/long averages
	* First year: use/calculate from current years value
	* Second year: use the prior year's values
	* Third year+: Use previous years, up to <current year> - 1
- `calc_veg_predictor_vals()`: Swap isotherm and temp-precip-correlation when calculating anomalies
- `calc_veg_predictor_vals()`: Use season precipitation averages when calculating rate of anomaly rather than driest season
- `calc_CONUS_vegcov_2025()`: Use annual temp-precip-correlation rather than anomaly variant, when calculating `zltCorPrTas`

- Result: Resulting values still do not sum to 1

N1ckP3rsl3y committed:
Correct dynamic vegetation calculation part 2

- Part 1: 2575cee8284edb812a4edfe677ce3c1db77aa483

- Reminder: Calculated dynamic vegetation values do not sum to 1, meaning incorrect calculations

- Remove division by 100 in final calculation of `calc_CONUS_vegcov_2025()`

- Result: Resulting values now sum to 1, implying correct calculations

N1ckP3rsl3y committed:
Remove repetitive parenthesis in `correlation_coefficient()`

- Remove repetitive parenthesis in the following line
	* sqrt((xSquSum * ySquSum)) --> sqrt(xSquSum * ySquSum)

dschlaep committed:
update_veg_yearly() now incorporates also second simulation year

- previously, update_veg_yearly() skipped (1) the calculation of annual predictors for the second simulation year and (2) updating dynamic vegetation and soil temperature boundary condition
- now, the calculation of annual predictors is done for every year (including the second simulation year)
- now, only the update of dynamic vegetation and soil temperature boundary condition are skipped for the second simulation year (because they are identical to the first year)

dschlaep committed:
Dynamic vegetation is now providing annual estimates of cover

- update_veg_yearly() now updates simulation cover values by plant functional type annually with values estimated by calc_CONUS_vegcov_2025()

dschlaep committed:
Fix equations in calc_CONUS_vegcov_2025()

- the model for ecoregion classification now uses long-term average predictors (instead and incorrectly values of a specific year)
- the model of CONUS-wide bare ground cover is now fixed
- the model of the proportion of total tree that is needle-leaved in forests is now fixed

dschlaep committed:
Enable dynamic vegetation for spinup

- Dynamic vegetation now also works with spinup
- Years for dynamic vegetation are now counted sequentially starting with spinup and continuing through the simulation period
- WaterBalanceFixtureTest.WaterBalanceWithVegetationFromClimate2 (test dynamic vegetation) now includes spinup

dschlaep committed:
Fix calculation of annual mean temperature by calc_yearly_hist_vals()

- calc_yearly_hist_vals() now calculates annual mean temperature as the annual average of daily values
- previously, it calculated the mean of monthly averages which tends to have a small cold bias in climates if May-Oct (184 days) are warmer than Nov-Apr (181 or 182 days)

dschlaep committed:
Update NEWS
